### PR TITLE
[8.x] Add option to release unique job locks before processing

### PR DESCRIPTION
**Pull request by paras-malhotra**
Based on our discussion at https://github.com/laravel/framework/pull/35042, there may be use cases where it is desirable to release unique job locks before the job starts processing (to avoid race conditions between job completion and unlocking). This PR allows the option to release the unique job lock before processing if the `$uniqueUntilStart` is set to true like so:

```php
class ExampleJob implements ShouldQueue, ShouldBeUnique
{
    use InteractsWithQueue, Queueable, Dispatchable;

    public $uniqueUntilStart = true;
}
```

This is backwards-compatible. If the job raises an error, we don't attempt to re-acquire the lock and may be duplicate enqueued. This should be fine as anyone who wants to use this feature would need to make sure that the job is idempotent because a duplicate job dispatch can happen while the job is processing. This is similar to [Sidekiq's unique_until start feature](https://github.com/mperham/sidekiq/wiki/Ent-Unique-Jobs#unlock-policy).

### CODE REVIEW

File: src/Illuminate/Queue/CallQueuedHandler.php [hunk 1]
```diff
@@ -59,6 +59,10 @@ public function call(Job $job, array $data)
             return $this->handleModelNotFound($job, $e);
         }
 
+        if ($command->uniqueUntilStart ?? false) {
+            $this->ensureUniqueJobLockIsReleased($command);
```
Comment (taylorotwell) [inline src/Illuminate/Queue/CallQueuedHandler.php:63]: If we release the lock here and then also release it when the job is finished processing then don't we risk releasing the lock for a job that was queued after this one started processing but before it finishes processing? Then we will release the lock and there could be duplicate jobs on the queue that haven't started processing yet.

**Comment by paras-malhotra** [comment]
@taylorotwell I've added a variable to ensure the lock is released only once in the job lifecycle.

**Comment by taylorotwell** [comment]
Are we certain that the `failed` method is called on the same instance so that the variable's value is still in place?

**Comment by taylorotwell** [comment]
Got rid of property. Added interface:

```php
class SomeJob implements ShouldBeUniqueUntilProcessing
{

}
```

**Pull request merged**